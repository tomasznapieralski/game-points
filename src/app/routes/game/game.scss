.game {
  display: flex;
  height: 600px;
  border: 2px solid #885900;
  border-radius: 20px;
  background-color: rgba(37, 21, 1, 0.51);
  color: #ffffff;

  &__board {
    width: 560px;
  }

  &__menu {
    width: 340px;
  }
}
