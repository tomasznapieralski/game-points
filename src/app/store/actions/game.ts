import { Action } from 'redux';

import { CollectedItemInterface } from '../interfaces/item';
import { BoardSlotInterface } from '../interfaces/board';

const prefix = '[GAME]';

export const GAME_INIT = `${prefix} INIT`;

export const gameInit = (): Action => ({
  type: GAME_INIT,
});

export const GAME_SAVE_ITEMS = `${prefix} SAVE_ITEMS`;

export interface GameSaveItemsAction {
  type: string;
  items: CollectedItemInterface[];
}

export const gameSaveItems = (items: CollectedItemInterface[]): GameSaveItemsAction => ({
  type: GAME_SAVE_ITEMS,
  items,
});

export const GAME_SAVE_BOARD_SLOTS = `${prefix} SAVE_BOARD_SLOTS`;

export interface GameSaveBoardSlotsAction {
  type: string;
  boardSlots: BoardSlotInterface[];
}

export const gameSaveBoardSlots = (boardSlots: BoardSlotInterface[]): GameSaveBoardSlotsAction => ({
  type: GAME_SAVE_BOARD_SLOTS,
  boardSlots,
});

export const GAME_START = `${prefix} START`;
export const GAME_STOP = `${prefix} STOP`;

export const gameStart = (): Action => ({
  type: GAME_START,
});

export const gameStop = (): Action => ({
  type: GAME_STOP,
});

export const GAME_SCORE_ITEM = `${prefix} SCORE_ITEM`;

export interface GameScoreItemAction {
  type: string;
  id: number;
}

export const gameScoreItem = (id: number): GameScoreItemAction => ({
  type: GAME_SCORE_ITEM,
  id,
});

export const GAME_SHOW_ITEM = `${prefix} SHOW_ITEM`;

export interface GameShowItemAction {
  type: string;
  id: number;
}

export const gameShowItem = (id: number): GameShowItemAction => ({
  type: GAME_SHOW_ITEM,
  id,
});

export const GAME_HIDE_ITEM = `${prefix} HIDE_ITEM`;

export interface GameHideItemAction {
  type: string;
  id: number;
}

export const gameHideItem = (id: number): GameHideItemAction => ({
  type: GAME_HIDE_ITEM,
  id,
});

export const GAME_REMOVE_ITEM = `${prefix} REMOVE_ITEM`;

export interface GameRemoveItemAction {
  type: string;
  id: number;
}

export const gameRemoveItem = (id: number): GameRemoveItemAction => ({
  type: GAME_REMOVE_ITEM,
  id,
});
