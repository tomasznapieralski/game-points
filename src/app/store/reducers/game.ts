import { Action } from 'redux';

import { CollectedItemInterface, ItemInterface } from '../interfaces/item';
import { BoardSlotInterface } from '../interfaces/board';

import {
  GAME_SAVE_ITEMS,
  GameSaveItemsAction,
  GAME_SAVE_BOARD_SLOTS,
  GameSaveBoardSlotsAction,
  GAME_START,
  GAME_STOP,
} from '../actions/game';

import { ITEMS } from '../../constraints/items';

export interface GameReducerInterface {
  isGameRunning: boolean;
  items: ItemInterface[];
  collectedItems: CollectedItemInterface[];
  boardSlots: BoardSlotInterface[];
}

const defaultState: GameReducerInterface = {
  isGameRunning: false,
  items: [{
    icon: ITEMS.FIRE,
    basePoints: 200,
    bonusPoints: 50,
    bonusMultiplier: 2,
  }, {
    icon: ITEMS.FISH,
    basePoints: 150,
    bonusPoints: 100,
    bonusMultiplier: 3,
  }, {
    icon: ITEMS.LEMON,
    basePoints: 100,
    bonusPoints: 0,
    bonusMultiplier: 0,
  }, {
    icon: ITEMS.SKULL,
    basePoints: 50,
    bonusPoints: 0,
    bonusMultiplier: 0,
  }, {
    icon: ITEMS.SPIDER,
    basePoints: 25,
    bonusPoints: 1000,
    bonusMultiplier: 10,
  }, {
    icon: ITEMS.BIOHAZARD,
    basePoints: -250,
    bonusPoints: -50,
    bonusMultiplier: 3,
  }],
  collectedItems: [],
  boardSlots: [],
};

export default (state = defaultState, action: Action): GameReducerInterface => {
  switch (action.type) {
    case GAME_SAVE_ITEMS: {
      return {
        ...state,
        collectedItems: [...(action as GameSaveItemsAction).items],
      };
    }
    case GAME_SAVE_BOARD_SLOTS: {
      return {
        ...state,
        boardSlots: [...(action as GameSaveBoardSlotsAction).boardSlots],
      };
    }
    case GAME_START: {
      return {
        ...state,
        isGameRunning: true,
      };
    }
    case GAME_STOP: {
      return {
        ...state,
        isGameRunning: false,
      };
    }
    default: {
      return state;
    }
  }
};
