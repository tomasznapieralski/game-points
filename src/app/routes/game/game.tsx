import React, { useEffect } from 'react';
import { Dispatch, Action } from 'redux';
import { connect } from 'react-redux';

import Board from './components/board/board';
import Menu from './components/menu/menu';

import { AppStateInterface } from '../../store/reducers';

import { CollectedItemInterface } from '../../store/interfaces/item';
import { BoardSlotInterface } from '../../store/interfaces/board';

import {
  getCollectedItems,
  getBoardSlots,
  getIsGameRunning,
} from '../../store/selectors/game';

import {
  gameInit,
  gameScoreItem,
  gameShowItem,
  gameHideItem,
  gameStart,
  gameStop,
} from '../../store/actions/game';

import './game.scss';

interface PropsInterface {
  isGameRunning: boolean;
  collectedItems: CollectedItemInterface[];
  boardSlots: BoardSlotInterface[];
  gameInitAction: () => Action;
  gameStartAction: () => Action;
  gameStopAction: () => Action;
  gameScoreItemAction: (id: number) => Action;
  gameShowItemAction: (id: number) => Action;
  gameHideItemAction: (id: number) => Action;
}

export const Game: React.FC<PropsInterface> = ({
  isGameRunning,
  collectedItems,
  boardSlots,
  gameInitAction,
  gameStartAction,
  gameStopAction,
  gameScoreItemAction,
  gameShowItemAction,
  gameHideItemAction,
}) => {
  useEffect(() => {
    gameInitAction();
  }, [gameInitAction]);

  return (
    <div className="game">
      <div className="game__board">
        <Board
          boardSlots={boardSlots}
          handleBoardItemClick={gameScoreItemAction}
          handleBoardItemInit={gameShowItemAction}
          handleBoardItemTimeout={gameHideItemAction}
        />
      </div>
      <div className="game__menu">
        <Menu
          collectedItems={collectedItems}
          isGameRunning={isGameRunning}
          startGameHandler={gameStartAction}
          stopGameHandler={gameStopAction}
        />
      </div>
    </div>
  );
}

const mapStateToProps = (state: AppStateInterface) => ({
  isGameRunning: getIsGameRunning(state),
  collectedItems: getCollectedItems(state),
  boardSlots: getBoardSlots(state),
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  gameInitAction: () => dispatch(gameInit()),
  gameStartAction: () => dispatch(gameStart()),
  gameStopAction: () => dispatch(gameStop()),
  gameScoreItemAction: (id: number) => dispatch(gameScoreItem(id)),
  gameShowItemAction: (id: number) => dispatch(gameShowItem(id)),
  gameHideItemAction: (id: number) => dispatch(gameHideItem(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Game);
