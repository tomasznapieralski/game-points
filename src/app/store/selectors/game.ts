import { AppStateInterface } from '../reducers/index';

export const getItems = (state: AppStateInterface) => state.game.items;

export const getCollectedItems = (state: AppStateInterface) => state.game.collectedItems;

export const getBoardSlots = (state: AppStateInterface) => state.game.boardSlots;

export const getIsGameRunning = (state: AppStateInterface) => state.game.isGameRunning;
