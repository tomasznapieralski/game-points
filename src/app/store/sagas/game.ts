import { takeEvery, select, put, take, race, call, delay } from 'redux-saga/effects';

import {
  GAME_INIT,
  GAME_START,
  GAME_STOP,
  gameSaveItems,
  gameSaveBoardSlots,
  GAME_SCORE_ITEM,
  GameScoreItemAction,
  GAME_SHOW_ITEM,
  GameShowItemAction,
  GAME_REMOVE_ITEM,
  GameRemoveItemAction,
  gameRemoveItem,
  GAME_HIDE_ITEM,
  GameHideItemAction,
  gameHideItem,
} from '../actions/game';

import { getItems, getBoardSlots, getCollectedItems } from '../selectors/game';

import { ItemInterface, CollectedItemInterface } from '../interfaces/item';
import { BoardSlotInterface } from '../interfaces/board';

import { BOARD_DIMENSIONS } from '../../constraints/board';

import { getRandomInt } from '../../utils/random';

function* setupItems() {
  yield takeEvery(GAME_INIT, function* () {
    const items: ItemInterface[] = yield select(getItems);
    const collectedItems = items.map((item) => {
      return {
        item,
        quantity: 0,
      } as CollectedItemInterface;
    });

    yield put(gameSaveItems(collectedItems));
  });
}

function* setupBoard() {
  yield takeEvery(GAME_INIT, function* () {
    const boardSlots: BoardSlotInterface[] = [];

    for (let row = 1; row <= BOARD_DIMENSIONS.ROWS; row++) {
      for (let col = 1; col <= BOARD_DIMENSIONS.COLS; col++) {
        boardSlots.push({
          row,
          col,
          id: Number(`${row}${col}`),
          isHidden: true,
          item: null,
        } as BoardSlotInterface);
      }
    }

    yield put(gameSaveBoardSlots(boardSlots));
  });
}

function* showItem() {
  yield takeEvery(GAME_SHOW_ITEM, function* (action: GameShowItemAction) {
    const boardSlots: BoardSlotInterface[] = yield select(getBoardSlots);
    const boardSlotIndex = boardSlots.findIndex((boardSlot) => boardSlot.id === action.id);
    const newBoardSlots = [...boardSlots];

    newBoardSlots[boardSlotIndex] = {
      ...newBoardSlots[boardSlotIndex],
      isHidden: false,
    };

    yield put(gameSaveBoardSlots(newBoardSlots));
  });
}

function* hideItem() {
  yield takeEvery(GAME_HIDE_ITEM, function* (action: GameHideItemAction) {
    const boardSlots: BoardSlotInterface[] = yield select(getBoardSlots);
    const boardSlotIndex = boardSlots.findIndex((boardSlot) => boardSlot.id === action.id);
    const newBoardSlots = [...boardSlots];

    newBoardSlots[boardSlotIndex] = {
      ...newBoardSlots[boardSlotIndex],
      isHidden: true,
    };

    yield put(gameSaveBoardSlots(newBoardSlots));
    yield put(gameRemoveItem(action.id));
  });
}

function* scoreItem() {
  yield takeEvery(GAME_SCORE_ITEM, function* (action: GameScoreItemAction) {
    const boardSlots: BoardSlotInterface[] = yield select(getBoardSlots);
    const collectedItems: CollectedItemInterface[] = yield select(getCollectedItems);
    const boardSlotIndex = boardSlots.findIndex((boardSlot) => boardSlot.id === action.id);
    const newBoardSlots = [...boardSlots];
    const boardSlotItem = newBoardSlots[boardSlotIndex].item;
    const collectedItemIndex = collectedItems.findIndex(
      (collectedItem) => collectedItem.item.icon === (boardSlotItem && boardSlotItem.icon)
    );
    const newCollectedItems = [...collectedItems];

    newBoardSlots[boardSlotIndex] = {
      ...newBoardSlots[boardSlotIndex],
      isHidden: true,
    };

    newCollectedItems[collectedItemIndex] = {
      ...newCollectedItems[collectedItemIndex],
      quantity: newCollectedItems[collectedItemIndex].quantity + 1,
    };

    yield put(gameSaveBoardSlots(newBoardSlots));
    yield put(gameSaveItems(newCollectedItems));
    yield put(gameRemoveItem(action.id));
  });
}

function* removeItem() {
  yield takeEvery(GAME_REMOVE_ITEM, function* (action: GameRemoveItemAction) {
    yield delay(300);

    const boardSlots: BoardSlotInterface[] = yield select(getBoardSlots);
    const boardSlotIndex = boardSlots.findIndex((boardSlot) => boardSlot.id === action.id);
    const newBoardSlots = [...boardSlots];

    newBoardSlots[boardSlotIndex] = {
      ...newBoardSlots[boardSlotIndex],
      item: null,
    };

    yield put(gameSaveBoardSlots(newBoardSlots));
  });
}

function* spawnItems() {
  while (true) {
    yield take(GAME_START);
    yield race({
      task: call(function* () {
        while (true) {
          const items: ItemInterface[] = yield select(getItems);
          const boardSlots: BoardSlotInterface[] = yield select(getBoardSlots);
          const availableBoardSlots = boardSlots.filter((boardSlot) => boardSlot.item === null);

          if (availableBoardSlots.length > 0) {
            const newBoardSlots = [...boardSlots];
            const randomSlotIndex = getRandomInt(0, availableBoardSlots.length - 1);
            const randomSlot = availableBoardSlots[randomSlotIndex];
            const newBoardSlotIndex = newBoardSlots.findIndex((boardSlot) =>
              boardSlot.id === randomSlot.id
            );
            const randomItem = getRandomInt(0, items.length - 1);

            newBoardSlots[newBoardSlotIndex] = {
              ...newBoardSlots[newBoardSlotIndex],
              isHidden: false,
              item: items[randomItem],
            };

            yield put(gameSaveBoardSlots(newBoardSlots));
          }

          yield delay(200);
        }
      }),
      cancel: take(GAME_STOP),
    })
  }
}

function* hideAllItems() {
  yield takeEvery(GAME_STOP, function* () {
    const boardSlots: BoardSlotInterface[] = yield select(getBoardSlots);

    for (let { isHidden, item, id } of boardSlots) {
      if (!isHidden && item) {
        yield put(gameHideItem(id));
      }
    }
  });
}

function* resetScores() {
  yield takeEvery(GAME_START, function* () {
    const collectedItems: CollectedItemInterface[] = yield select(getCollectedItems);
    const newCollectedItems = collectedItems.map((collectedItem) => ({
      ...collectedItem,
      quantity: 0,
    }));

    yield put(gameSaveItems(newCollectedItems));
  });
}

export default [
  setupItems(),
  setupBoard(),
  showItem(),
  hideItem(),
  scoreItem(),
  removeItem(),
  spawnItems(),
  hideAllItems(),
  resetScores(),
];
