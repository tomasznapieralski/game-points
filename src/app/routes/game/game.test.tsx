import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup } from '@testing-library/react';

import { Game } from './game';

afterEach(cleanup);

test('Snapshot Correct Component', () => {
  const tree = renderer
    .create(
      <Game
        isGameRunning={false}
        collectedItems={[]}
        boardSlots={[]}
        gameInitAction={jest.fn()}
        gameStartAction={jest.fn()}
        gameStopAction={jest.fn()}
        gameScoreItemAction={jest.fn()}
        gameShowItemAction={jest.fn()}
        gameHideItemAction={jest.fn()}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
