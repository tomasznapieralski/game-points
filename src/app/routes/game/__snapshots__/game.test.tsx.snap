// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Correct Component 1`] = `
<div
  className="game"
>
  <div
    className="game__board"
  >
    <div
      className="board"
    >
      <div
        className="board__title"
      >
        <img
          alt=""
          className="board__logo"
          src="logo.png"
        />
        <div
          className="board__subtitle"
        >
          Points
        </div>
      </div>
      <div
        className="board__playground"
      />
    </div>
  </div>
  <div
    className="game__menu"
  >
    <div
      className="menu"
    >
      <div
        className="menu__title"
      >
        Player Items
      </div>
      <div
        className="menu__space"
      />
      <div
        className="menu__footer"
      >
        <ul
          className="scores-list"
        >
          <li
            className="scores-list__item"
          >
            <div>
              Base points:
            </div>
            <div>
              0
            </div>
          </li>
          <li
            className="scores-list__item"
          >
            <div>
              Bonus points:
            </div>
            <div>
              0
            </div>
          </li>
          <li
            className="scores-list__item"
          >
            <div>
              Total points:
            </div>
            <div>
              0
            </div>
          </li>
        </ul>
        <button
          className="menu__footer-button menu__footer-button--green"
          onClick={[MockFunction]}
        >
          Start Game
        </button>
      </div>
    </div>
  </div>
</div>
`;
